### PR TITLE
feat: pie-chart support negative value

### PR DESCRIPTION
**Pull request by ssthouse**
## Brief Information

This pull request is in the type of:

- [x] bug fixing
- [ ] new feature
- [ ] others

### What does this PR do?

Support minus value in Pie Chart.

### Fixed issues

- #15055

### Before: What was the problem?

the minus value slice fillup the entire pie chart;

### After: How is it fixed in this PR?

image

## Misc

- [ ] The API has been changed (apache/echarts-doc#xxx).
- [ ] This PR depends on ZRender changes (ecomfe/zrender#xxx).

### Related test cases or examples to use the new APIs

NA.

add test file: `pie-minus-value.html` in test folder;

### Merging options

- [ ] Please squash the commits into a single one when merge.

### CODE REVIEW

File: src/chart/pie/PieSeries.ts [hunk 1]
```diff
@@ -170,7 +170,13 @@ class PieSeriesModel extends SeriesModel<PieSeriesOption> {
...
         const valueList: number[] = [];
         data.each(data.mapDimension('value'), function (value: number) {
-            valueList.push(value);
+            // handle minue value condition: teat minus value as 0 when calculation percent
+            if (!isNaN(value) && value < 0) {
+                valueList.push(0);
+            }
+            else {
+                valueList.push(value);
+            }
```
Comment (Ovilia) [inline src/chart/pie/PieSeries.ts:179]: This could be simplified to be `valueList.push(isNaN(value) ? value : Math.max(value, 0));`

**Review by Ovilia — CHANGES_REQUESTED** [review]

File: src/chart/pie/PieSeries.ts (hunk 1 above)
Comment (ssthouse) [reply src/chart/pie/PieSeries.ts]: fixed

File: src/chart/pie/pieLayout.ts [hunk 2]
```diff
@@ -72,8 +72,13 @@ export default function pieLayout(
         data.each(valueDim, function (value: number) {
             !isNaN(value) && validDataCount++;
         });
+        let sum = 0;
+        data.each(valueDim, function (value: number) {
+            if (!isNaN(value) && value > 0) {
+                sum += value;
+            }
+        });
```
[the remark this answers is not in the document]
Comment (ssthouse) [reply src/chart/pie/pieLayout.ts]: fixed

**Comment by Ovilia** [comment]
@ssthouse Could you try to filter negative data with pie series in `processor/dataFilter.ts`? If it works, you don't need to change pieView and it would be simpler. Please be careful when checking the legend of negative data.

**Comment by ssthouse** [comment]
> @ssthouse Could you try to filter negative data with pie series in `processor/dataFilter.ts`? If it works, you don't need to change pieView and it would be simpler. Please be careful when checking the legend of negative data.

I'll try~

**Comment by ssthouse** [comment]
@Ovilia hi, do you mean adding filter logic like this?

image


So I can discard the changed I've made in `src/chart/pie/pieLayout.ts` and `src/chart/pie/PieSeries.ts`

**Comment by pissang** [comment]
@ssthouse Should register a new processor to filter the negative data. See https://github.com/apache/echarts/blob/master/src/chart/pie/install.ts#L36 about how to register a processor

**Comment by ssthouse** [comment]
@pissang hi, I create a new filter: `negativeDataFilter.ts` and register it in `pie/install.ts`, please check~

**Comment by plainheart** [comment]
It seems this will break the case in which all the number is negative. Besides, if the negative value was considered as `0`, how does the tooltip show? Should it be `0`(as the pie shows) or the original negative value like `-1548`?

**Comment by ssthouse** [comment]
> It seems this will break the case in which all the number is negative. Besides, if the negative value was considered as `0`, how does the tooltip show? Should it be `0`(as the pie shows) or the original negative value like `-1548`?

@plainheart hi, In my previous edition(see the first two commits of this PR).
the negative value is considered as `0` when calculating pie slices. And the tooltip shows the negative value label; like below:

image

---

according to @Ovilia and @pissang's suggestion, I moved the filter logic to `processor/negativeDataFilter`( see the 3rd commit of this PR). 
and in this edition, the negative value will only be shown in legend and not in pie slices. like below:
image

I'm not sure which result is expected since they're both reasonable in some way，hope you guys can give me a conclusion😂

**Comment by Ovilia** [comment]
I think it's a little better if not show in the tooltip but show in legend.

File: src/processor/negativeDataFilter.ts [hunk 3]
```diff
@@ -0,0 +1,44 @@
+import { StageHandler } from '../util/types';
```
Comment (Ovilia) [inline src/processor/negativeDataFilter.ts:1]: This should appear after the license header.

File: src/chart/pie/pieLayout.ts [hunk 4]
```diff
@@ -30,9 +30,9 @@ const RADIAN = Math.PI / 180;
 function getViewRect(seriesModel: PieSeriesModel, api: ExtensionAPI) {
     return layout.getLayoutRect(
         seriesModel.getBoxLayoutParams(), {
-            width: api.getWidth(),
-            height: api.getHeight()
-        }
+        width: api.getWidth(),
+        height: api.getHeight()
+    }
```
Comment (Ovilia) [inline src/chart/pie/pieLayout.ts:35]: I think these three lines (or perhaps this file?) should not be changed.

File: test/pie-minus-value.html [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+                    },
+                    series: [
+                        {
+                            name: "访问来源",
+                            type: "pie",
+                            radius: "55%",
+                            center: ["50%", "60%"],
+                            selectedMode: "single",
+                            data: [
+                                { value: 335, name: "直接访问" },
```
Comment (Ovilia) [inline test/pie-minus-value.html:80]: As @plainheart suggested, I think we should add a test case that all data are negative.

**Review by Ovilia — CHANGES_REQUESTED** [review]

File: src/chart/pie/pieLayout.ts (hunk 4 above)
Comment (ssthouse) [reply src/chart/pie/pieLayout.ts]: fixed

File: src/processor/negativeDataFilter.ts (hunk 3 above)
Comment (ssthouse) [reply src/processor/negativeDataFilter.ts]: fixed

File: test/pie-minus-value.html (hunk 5 above)
Comment (ssthouse) [reply test/pie-minus-value.html]: When all the data are negative, the pie chart is empty.
I'm not sure if this is the expected result 🐒

image

Comment (plainheart) [reply test/pie-minus-value.html]: > When all the data are negative, the pie chart is empty.
> I'm not sure if this is the expected result 🐒

At least, it's unexpected for me. This will break the previous behavior. And I think it's meaningful when all value is negative. In that case, considering them as `0` may be not a good way.

Comment (Ovilia) [reply test/pie-minus-value.html]: By default (if [series-pie.stillShowZeroSum](https://echarts.apache.org/zh/option.html#series-pie.stillShowZeroSum) is not set to be `false`), when all data is negative, it should behave the same as when all data is zero.

image

Comment (Ovilia) [reply test/pie-minus-value.html]: If the canvas is blank, it looks like there is an error to the users. So perhaps we could display an empty light gray circle when no data (sum is zero or all is `'-'` or negative). We should also provide options to configure the color.
The legend color doesn't look right to me. They should have different colors for different series.

Comment (ssthouse) [reply test/pie-minus-value.html]: @Ovilia this sounds reasonable, I'll give it a try~
>  display an empty light gray circle when no data

For the color issue, all three legends' are in blue because they are all filtered by `negativeDataFilter`. And the filtered legend's default color is blue

**Comment by ssthouse** [comment]
@Ovilia Hi~
I added an empty light gray circle when the data count is 0 or all data are filtered (by `negativeDataFilter.ts`). It looks like this:
wecom-temp-441629a7e1e40f5ef839a4027ad86d34

---
This is what it looks when data are dynamically changed to be all negative. 
81cdeb2b76890aa7b2440d9f220a7941

---
This is what it looks when two data are negative and the other two are 0 (so they are summed to be 0):

Jun-15-2021 22-39-14

**Comment by pissang** [comment]
@ssthouse Awesome! Can we have the ability to configure if display this empty circle and the style. Perhaps the options can be:

```
showEmptyCircle: true,
emptyCircleStyle: {
  // same with itemStyle
}
```

**Comment by ssthouse** [comment]
> @ssthouse Awesome! Can we have the ability to configure if display this empty circle and the style. Perhaps the options can be:
> 
> ```js
> showEmptyCircle: true,
> emptyCircleStyle: {
>   // same with itemStyle
> }
> ```

@pissang Yes, sure~

I have a small question that may need your help. 🙏 

About the option `emptyCircleStyle`. Currently, I implement the light-gray circle by creating a simple `graphic.Sector`, and the type of itemStyle (`ItemStyleOption`) seems can not be directly applied to `Sector`.

I checked the sector's API and found it accepts `PathStyleProps`, maybe we can expose `emptyCircleStyle: PathStyleProps` ?

image

**Comment by pissang** [comment]
@ssthouse Here is a common way to set style from the option:

```ts
sector.useStyle(seriesModel.getModel('emptyCircleStyle').getItemStyle());
```

Some note:

+ Need to use `useStyle` here so the style properties can be reset to default when developers didn't set it in the new options. 
+ `getItemStyle` get the full mapped style object from the model. It includes fill and stroke. Meanwhile, there are other two methods that can get a mapped style object. `getLineStyle` which only includes stroke. `getAreaStyle` which only includes fill.

**Comment by ssthouse** [comment]
@pissang Thanks for your suggestions, really helpful!

I have added the config options, please check~

image

File: src/processor/negativeDataFilter.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+export default function negativeDataFilter(seriesType: string): StageHandler {
+    return {
+        seriesType: seriesType,
+        reset: function (seriesModel, ecModel) {
+            const legendModels = ecModel.findComponents({
+                mainType: 'legend'
+            });
+            if (!legendModels || !legendModels.length) {
```
Comment (pissang) [inline src/processor/negativeDataFilter.ts:29]: I think it's not necessary to check if there is `legend`?

File: src/chart/pie/PieView.ts [hunk 7]
```diff
@@ -242,6 +243,29 @@ class PieView extends ChartView {
             }
         }
 
+        // remove empty-circle(which is not instance of PiePiece) if it exists
+        group.children().forEach(child => {
+            if (!(child instanceof PiePiece)) {
```
Comment (pissang) [inline src/chart/pie/PieView.ts:248]: It's not a good way to check if it's the empty-circle element by checking it's not a piece of the pie. It may have issues if we add other elements that are not a piece of the pie further.

A better way is to save this empty-circle element as a private property of PieView. It's simpler and more robust.

**Review by pissang — CHANGES_REQUESTED** [review]

File: src/processor/negativeDataFilter.ts (hunk 6 above)
Comment (ssthouse) [reply src/processor/negativeDataFilter.ts]: fixed

File: src/chart/pie/PieView.ts (hunk 7 above)
Comment (ssthouse) [reply src/chart/pie/PieView.ts]: fixed

**Review by pissang — DISMISSED** [review]

**Comment by pissang** [comment]
LGTM now

File: src/chart/pie/PieView.ts [hunk 8]
```diff
@@ -242,6 +244,28 @@ class PieView extends ChartView {
...
+            const sector = new graphic.Sector({
+                shape: {
+                    cx,
+                    cy,
+                    r0,
+                    r,
+                    startAngle: 0,
+                    endAngle: Math.PI * 2
+                }
+            });
```
Comment (plainheart) [inline src/chart/pie/PieView.ts:263]: The sector has a default value for `startAngle` and `endAngle`, so this can be simplified as
```js
const sector = new graphic.Sector({
    shape: getBasicPieLayout(seriesModel, api)
});
```

**Review by plainheart — COMMENTED** [review]
Thanks for the work. It should be okay generally. But I can't keep confident about the final process: _**filter the negative values**_.

Comparing to current behavior, this result looks more expected for me,



That is, leave the negative value there and its percentage is _**0**_.

And if all values are negative, display them as the same as the positive ones. (BTW, this is also the behavior before this PR's changes.)

File: src/chart/pie/PieView.ts (hunk 8 above)
Comment (ssthouse) [reply src/chart/pie/PieView.ts]: fixed

**Review by plainheart — APPROVED** [review]

**Comment by pissang** [comment]
@plainheart About the negative value filtering. I can accept it mainly because highcharts is using the same strategy. But I'm also not sure if there are developers using this `bug` as a `feature` to get the result they want. If so this change may break it and we need an extra option to not filter this negative values.

**Comment by ssthouse** [comment]
@pissang Should I add this config now or we wait for developers‘ feedback after release?

**Comment by pissang** [comment]
@ssthouse I think we can wait for the developers' feedback. So I'm going to merge it.

**Review by Ovilia — APPROVED** [review]
LGTM

**Comment by yangkai-bj** [comment]
在Pie中强制过滤负数,是否简单粗暴了一些,在饼图和圆环图中负数确实导致图形绘制的问题,但玫瑰图可以真实反映负数,是否可以将PIE中"是否过滤负值"作为一个用户选项.
日常涉及规模数据的时候一般都是非负数,使用Pie绘制很好理解,但涉及增量的占比,负增数据无法避免,饼图和圆环图都无法绘制,往往需要对原始数据进行变形(将所有值+abs(最小负值))后绘制,绘制后的图表让人很难理解,5.1.2以前的玫瑰图(roseType: "area")可以真实体现原始数据中的负数.

**Comment by yangkai-bj** [comment]
TEST10 (2)

**Comment by pissang** [comment]
@yangkai-bj 谢谢建议，欢迎 PR 增加该配置
